WIP: HDR rendering and output support

Filoppi opened the pull request:
Pls test in many games for both results (how good it looks) and bugs.
Most games are bugfree with this from my tests, and look amazing.

Showcase:
https://vimeo.com/1077587822
https://vimeo.com/1077588151
https://vimeo.com/1077588981
https://vimeo.com/1077587655
https://vimeo.com/1077587529
https://vimeo.com/1077588807

Filoppi committed:
WIP: HDR

Filoppi committed:
Fix shader build

Filoppi committed:
Fix shader

Filoppi committed:
Fix textures not reloading when toggling the button live

foxsevent commented:
Shadows seems to be bugged in TLOS A New beginning with HDR on under a specific angles, so a minor bug?
![image](https://github.com/user-attachments/assets/ae6a858d-937d-4adc-804c-8b9addc5a3fb)
![image](https://github.com/user-attachments/assets/4e827306-d64c-4525-8c05-bfeead0d56f7)
GS dump if that helps.
[The Legend of Spyro - A New Beginning_SLUS-21372_20250421011102.gs.zip](https://github.com/user-attachments/files/19827435/The.Legend.of.Spyro.-.A.New.Beginning_SLUS-21372_20250421011102.gs.zip)

Filoppi commented:
@foxsevent thanks. testing is happening on discord, if you can, it'd help to have you there as well.

Filoppi committed:
More

Filoppi committed:
Fix mac

Filoppi committed:
Remove old hacky HDR implementation

Filoppi committed:
Remove unnecessary force SW blends

Filoppi committed:
Hide dithering in HDR

Filoppi committed:
Remove hacky DX11 blends

Filoppi committed:
Fix HDR render api toggle

Filoppi committed:
Improve clamping/round for shuffles

Filoppi committed:
Shader improvements

Filoppi committed:
Revert "Remove unnecessary force SW blends"

This reverts commit 3199dea03d0b20f32d0c3492bf2192e9df6847fd.

Margen67 commented on bin/resources/shaders/common/fxaa.fx:
```diff
@@ -153,21 +140,36 @@ float4 PreGammaPass(float4 color)
+	rgbyM = FxaaEncode(rgbyM);
 	#define lumaM rgbyM.w
-
+	
```
Whitespace

Filoppi committed:
Add HDR custom textures support

Filoppi committed:
Minor polish

Filoppi committed:
Always quantize HDR textures alphas (fixes Jak games shadows)

Filoppi committed:
Add linear tfx sampler

Filoppi committed:
Specify default filter

Filoppi committed:
Add new flag to HW blends combinations to specify whichs ones are subtractive, because in HDR we use R16G16B16A16F textures so if we did blends in HW, they could end up with values below zero, thus we can force them to run in SW (and clip values below 0) in HDR.

Filoppi committed:
Add multiple HDR modes for quick compatibility testing
Polish code

Filoppi committed:
WIP Stretch textures

Filoppi committed:
More WIP

Filoppi committed:
Grey out dithering (instead of hiding it) in HDR

Filoppi committed:
Make sure CopyRect() doesn't crash if it's reading or writing to out of range coordinates

Filoppi committed:
Fix negative blend flag

Filoppi committed:
Lower max paper white to 400 as 500 could go beyond the peak brightness

Filoppi committed:
Fix LUT extrapolation reading the size from the wrong texture

Filoppi committed:
Polish

Filoppi committed:
WIP double shuffle detection (e.g. swap g<->a or r<->g without any other alterations to the color in the processs), for HDR we don't want to clip it to 1 (or 255) as it'd lose some scene detail

Filoppi committed:
Fix blend flags properly

Filoppi committed:
Conflict fixes

Filoppi committed:
Explain texture age

Filoppi committed:
Simplify hazard copy intersection handling

Filoppi committed:
Improve LUT extrapolation

Filoppi committed:
Fix FSMASK decimals

Filoppi committed:
Polish

Filoppi committed:
Clamp alpha in HDR for extra safety, some textures accumulate every frame (e.g. NFS Underground 2 bloom/trail) and are never cleared

Filoppi committed:
Fix max brightness

Filoppi committed:
Fix uint mismatch warning

Filoppi committed:
make DX shader compiler also write log on disc when there's just warnings, and fix the shader error name being the shader model

Filoppi committed:
Improving float to int rounding

Filoppi committed:
Fix shader error printing out the wrong name

Filoppi committed:
Split up HDR into 4 different presets (including "Off", which is identical to the old default behaviour).
Proper HDR is currently only available in DX11/12, but these have barrier problems with SW blends and thus produce many random glitches.
Press "L" to swap between the HDR modes (temporary code).
HDR is currently broken anyway, there's probably some dumb mistake in one of the last changes, no time to check now.

Filoppi committed:
Some HDR fixes

Filoppi committed:
Fix HDR peak brightness not applying correctly

Filoppi committed:
Add SDR->HDR hue restoration and improve the HDR in SDR look (by giving it more range)

Filoppi committed:
Better support for Mac  OS

RinMaru commented:
Would be really nice to have this. PerceptualHDR on Dolphin really makes colors pop. I guess theres no more builds of this for testing anymore? Or would it make any sense testing it now with how behind it is?
